src/middleware/open-graph/index.js

```diff
@@ -0,0 +1,207 @@
+// mac only. see: https://github.com/lovell/sharp/issues/2399#issuecomment-714381300
+// process.env.PANGOCAIRO_BACKEND = 'fontconfig';
+process.env.FONTCONFIG_PATH = 'fonts';
+
+const path = require('path');
+const got = require('got');
+const bytes = require('bytes');
+const sharp = require('sharp');
+const LRU = require('lru-cache');
+const FontsProcessor = require('./fonts');
+const { npmIndex } = require('../../lib/algolia');
+
+const API_HOST = 'https://data.jsdelivr.com';
+const LOGO_MAX_SIZE = 2 ** 20; // 1MiB
+
+const cache = new LRU({ max: 1000, maxAge: 24 * 60 * 60 * 1000 });
+const http = got.extend({ cache });
+
+const fontsProcessor = new FontsProcessor();
+fontsProcessor.addFont('Lexend Regular', path.resolve(__dirname, '../../../fonts/Lexend-Regular.ttf'));
+fontsProcessor.addFont('Lexend SemiBold', path.resolve(__dirname, '../../../fonts/Lexend-SemiBold.ttf'));
+
+const fetchStats = async (name, type = 'npm', period = 'month') => {
+	return http.get(`${API_HOST}/v1/package/${type}/${name}/stats/date/${period}`).json();
+};
+
+const fetchLogo = async (url) => {
+	return new Promise((resolve, reject) => {
+		let stream = http.stream.get(url);
+
+		let size = 0;
+		let mime = 'image/png';
+		let logo = [];
+
+		stream.on('end', () => {
+			resolve(`data:${mime};base64,${Buffer.concat(logo).toString('base64')}`);
+		});
+
+		stream.on('error', reject);
+
+		stream.on('response', (res) => {
+			if (res.headers['content-length'] && res.headers['content-length'] > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			mime = res.headers['content-type'] ?? 'image/png';
+		});
+
+		stream.on('data', (chunk) => {
+			size += chunk.length;
+
+			if (size > LOGO_MAX_SIZE) {
+				return stream.destroy(new Error('logo size limit exited'));
+			}
+
+			logo.push(chunk);
+		});
+	});
+};
+
+/**
+ * @param {string} input
+ * @param {string} fontFamily
+ * @param {number} fontSize
+ * @param {number} maxWidth
+ * @return {{width: number, text: string}}
+ */
+const truncateString = (input, fontFamily, fontSize, maxWidth) => {
+	let width = str => fontsProcessor.computeWidth(str, fontFamily, fontSize);
+	let truncate = (str) => {
+		let strWidth = width(str);
+		let dotsWidth = width('...');
+
+		if (strWidth <= maxWidth) {
+			return { text: str, width: strWidth };
+		}
+
+		while (strWidth > maxWidth - dotsWidth) {
+			str = str.substr(0, str.length - 1);
+			strWidth = width(str);
+		}
+
+		return { text: str + '...', width: strWidth + dotsWidth };
+	};
+
+	return truncate(input);
+};
+
+const processDescription = (description) => {
+	let maxLineWidth = 760;
+	let lineOffset = 263;
+	let lineHeight = 48;
+
+	let lines = fontsProcessor.wrap(description, 'Lexend Regular', 28, maxLineWidth);
+
+	if (lines.length > 2) {
+		lines = lines.slice(0, 2);
+		let lastLine = lines.pop();
+		let lastLineWidth = () => fontsProcessor.computeWidth(lastLine, 'Lexend Regular', 28);
```

Comment:
The svg size is 30 but I'm guessing that's compensated by the negative letter spacing?

Reply:
fixed